File: android/app/src/main/java/com/soobinpark/appcraft/readingbible/feature/reader/ReaderViewModel.kt
package com.soobinpark.appcraft.readingbible.feature.reader

import android.app.Application
import android.net.Uri
import android.os.Environment
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.soobinpark.appcraft.readingbible.data.preference.ReadingProgressPreferences
import com.soobinpark.appcraft.readingbible.data.repository.FileBibleRepository
import com.soobinpark.appcraft.readingbible.domain.model.BibleCatalog
import com.soobinpark.appcraft.readingbible.domain.model.BibleVerse
import com.soobinpark.appcraft.readingbible.domain.model.BibleVersion
import com.soobinpark.appcraft.readingbible.domain.model.ReadingProgress
import com.soobinpark.appcraft.readingbible.domain.model.VerseBookmark
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

data class ReaderUiState(
    val dataRoot: File = File(Environment.getExternalStorageDirectory(), "bible"),
    val versions: List<BibleVersion> = emptyList(),
    val selectedVersion: BibleVersion? = null,
    val comparisonVersion: BibleVersion? = null,
    val bookIndex: Int = 0,
    val chapter: Int = 1,
    val focusVerse: Int? = null,
    val verses: List<BibleVerse> = emptyList(),
    val comparisonVerses: List<BibleVerse> = emptyList(),
    val isLoading: Boolean = true,
    val loadingMessage: String = "성경 데이터를 준비하는 중입니다.",
    val message: String? = null,
)

class ReaderViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = FileBibleRepository(application)
    private val progressPreferences = ReadingProgressPreferences(application)
    private val _uiState = MutableStateFlow(ReaderUiState())
    val uiState: StateFlow<ReaderUiState> = _uiState.asStateFlow()
    private var dataFolderUri: Uri? = null
    private var localDataRoot: File? = null

    init {
        refresh()
    }

    fun setDataFolderUri(uri: Uri?) {
        if (dataFolderUri == uri) return
        dataFolderUri = uri
        localDataRoot = null
        refresh()
    }

    fun setDataFolder(
        uri: Uri?,
        localRoot: File?,
    ) {
        if (dataFolderUri == uri && localDataRoot == localRoot) return
        dataFolderUri = uri
        localDataRoot = localRoot
        _uiState.value = _uiState.value.copy(dataRoot = localRoot ?: File(Environment.getExternalStorageDirectory(), "bible"))
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            val root = _uiState.value.dataRoot
            _uiState.value = _uiState.value.copy(
                isLoading = true,
                loadingMessage = "데이터 폴더에서 BDF/LFA 역본을 찾는 중입니다.",
                message = null,
            )
            val versions = withContext(Dispatchers.IO) {
                dataFolderUri?.let { repository.scanVersions(it) }
                    ?.takeIf { it.isNotEmpty() }
                    ?: localDataRoot?.let { repository.scanVersions(it) }?.takeIf { it.isNotEmpty() }
                    ?: repository.scanVersions(root)
            }
            val savedProgress = progressPreferences.getProgress()
            val safeBookIndex = savedProgress.bookIndex.coerceIn(BibleCatalog.books.indices)
            val safeChapter = savedProgress.chapter.coerceIn(1, BibleCatalog.books[safeBookIndex].chapterCount)
            val selected = savedProgress.versionCode?.let { code ->
                versions.firstOrNull { it.code.equals(code, ignoreCase = true) }
            } ?: versions.firstOrNull()
            val comparison = if (savedProgress.hasSavedComparisonVersion) {
                savedProgress.comparisonVersionCode?.let { code ->
                    versions.firstOrNull {
                        it.code.equals(code, ignoreCase = true) && it.code != selected?.code
                    }
                }
            } else {
                versions.firstOrNull { it.code != selected?.code }
            }
            val book = BibleCatalog.books[safeBookIndex]
            _uiState.value = _uiState.value.copy(
                versions = versions,
                selectedVersion = selected,
                comparisonVersion = comparison,
                bookIndex = safeBookIndex,
                chapter = safeChapter,
                loadingMessage = selected?.let { "${it.code} ${book.koreanName} ${safeChapter}장을 여는 중입니다." }
                    ?: "사용 가능한 역본을 확인하는 중입니다.",
            )
            val (verses, comparisonVerses) = withContext(Dispatchers.IO) {
                val verses = selected?.let { repository.readChapter(it, book, safeChapter) }.orEmpty()
                val comparisonVerses = comparison?.let { repository.readChapter(it, book, safeChapter) }.orEmpty()
                verses to comparisonVerses
            }
            _uiState.value = _uiState.value.copy(
                verses = verses,
                comparisonVerses = comparisonVerses,
                isLoading = false,
                loadingMessage = "",
                message = when {
                    versions.isEmpty() -> "선택한 폴더 또는 기본 데이터 폴더에서 bdf/lfa 파일을 찾지 못했습니다."
                    selected != null && verses.isEmpty() -> "${selected.displayName} ${book.koreanName} ${safeChapter}장을 읽지 못했습니다. LFA/BDF 파일 형식 또는 손상 여부를 확인해 주세요."
                    else -> null
                },
            )
            _uiState.value.selectedVersion?.let { warmUpSelectedVersion(it) }
        }
    }

    fun selectVersion(version: BibleVersion) {
        viewModelScope.launch {
            val state = _uiState.value
            val (verses, comparison, comparisonVerses) = withContext(Dispatchers.IO) {
                val verses = repository.readChapter(version, BibleCatalog.books[state.bookIndex], state.chapter)
                val comparison = state.comparisonVersion?.takeUnless { it.code == version.code }
                val comparisonVerses = comparison?.let {
                    repository.readChapter(it, BibleCatalog.books[state.bookIndex], state.chapter)
                }.orEmpty()
                Triple(verses, comparison, comparisonVerses)
            }
            _uiState.value = state.copy(
                selectedVersion = version,
                comparisonVersion = comparison,
                verses = verses,
                comparisonVerses = comparisonVerses,
                message = if (verses.isEmpty()) {
                    val book = BibleCatalog.books[state.bookIndex]
                    "${version.displayName} ${book.koreanName} ${state.chapter}장을 읽지 못했습니다. LFA/BDF 파일 형식 또는 손상 여부를 확인해 주세요."
                } else {
                    null
                },
            )
            saveProgress(version, comparison, state.bookIndex, state.chapter)
            warmUpSelectedVersion(version)
        }
    }

    fun selectComparisonVersion(version: BibleVersion?) {
        viewModelScope.launch {
            val state = _uiState.value
            val comparison = version?.takeUnless { it.code == state.selectedVersion?.code }
            val verses = withContext(Dispatchers.IO) {
                comparison?.let {
                    repository.readChapter(it, BibleCatalog.books[state.bookIndex], state.chapter)
                }.orEmpty()
            }
            _uiState.value = state.copy(comparisonVersion = comparison, comparisonVerses = verses)
            saveProgress(state.selectedVersion, comparison, state.bookIndex, state.chapter)
        }
    }

    fun selectBookAndChapter(bookIndex: Int, chapter: Int) {
        viewModelScope.launch {
            val state = _uiState.value
            val safeBookIndex = bookIndex.coerceIn(BibleCatalog.books.indices)
            val safeChapter = chapter.coerceIn(1, BibleCatalog.books[safeBookIndex].chapterCount)
            val (verses, comparisonVerses) = withContext(Dispatchers.IO) {
                val verses = state.selectedVersion?.let {
                    repository.readChapter(it, BibleCatalog.books[safeBookIndex], safeChapter)
                }.orEmpty()
                val comparisonVerses = state.comparisonVersion?.let {
                    repository.readChapter(it, BibleCatalog.books[safeBookIndex], safeChapter)
                }.orEmpty()
                verses to comparisonVerses
            }
            _uiState.value = state.copy(
                bookIndex = safeBookIndex,
                chapter = safeChapter,
                focusVerse = null,
                verses = verses,
                comparisonVerses = comparisonVerses,
                message = if (state.selectedVersion != null && verses.isEmpty()) {
                    val book = BibleCatalog.books[safeBookIndex]
                    "${state.selectedVersion.displayName} ${book.koreanName} ${safeChapter}장을 읽지 못했습니다. LFA/BDF 파일 형식 또는 손상 여부를 확인해 주세요."
                } else {
                    null
                },
            )
            saveProgress(state.selectedVersion, state.comparisonVersion, safeBookIndex, safeChapter)
        }
    }

    fun openSearchResult(verse: BibleVerse) {
        viewModelScope.launch {
            val state = _uiState.value
            val versions = if (state.versions.isNotEmpty()) {
                state.versions
            } else {
                withContext(Dispatchers.IO) {
                    dataFolderUri?.let { repository.scanVersions(it) }
                        ?.takeIf { it.isNotEmpty() }
                        ?: localDataRoot?.let { repository.scanVersions(it) }?.takeIf { it.isNotEmpty() }
                        ?: repository.scanVersions(state.dataRoot)
                }
            }
            val safeBookIndex = verse.bookIndex.coerceIn(BibleCatalog.books.indices)
            val safeChapter = verse.chapter.coerceIn(1, BibleCatalog.books[safeBookIndex].chapterCount)
            val targetVersion = versions.firstOrNull { it.code.equals(verse.versionCode, ignoreCase = true) }
                ?: state.selectedVersion
                ?: versions.firstOrNull()
            val comparison = state.comparisonVersion
                ?.takeUnless { it.code == targetVersion?.code }
            val (verses, comparisonVerses) = withContext(Dispatchers.IO) {
                val targetVerses = targetVersion?.let {
                    repository.readChapter(it, BibleCatalog.books[safeBookIndex], safeChapter)
                }.orEmpty()
                val targetComparisonVerses = comparison?.let {
                    repository.readChapter(it, BibleCatalog.books[safeBookIndex], safeChapter)
                }.orEmpty()
                targetVerses to targetComparisonVerses
            }
            _uiState.value = state.copy(
                versions = versions,
                selectedVersion = targetVersion,
                comparisonVersion = comparison,
                bookIndex = safeBookIndex,
                chapter = safeChapter,
                focusVerse = verse.verse,
                verses = verses,
                comparisonVerses = comparisonVerses,
                isLoading = false,
                loadingMessage = "",
                message = if (targetVersion != null && verses.isEmpty()) {
                    val book = BibleCatalog.books[safeBookIndex]
                    "${targetVersion.displayName} ${book.koreanName} ${safeChapter}장을 읽지 못했습니다. LFA/BDF 파일 형식 또는 손상 여부를 확인해 주세요."
                } else {
                    null
                },
            )
            saveProgress(targetVersion, comparison, safeBookIndex, safeChapter)
            targetVersion?.let { warmUpSelectedVersion(it) }
        }
    }

    fun openBookmark(bookmark: VerseBookmark) {
        openSearchResult(
            BibleVerse(
                versionCode = bookmark.versionCode,
                bookIndex = bookmark.bookIndex,
                chapter = bookmark.chapter,
                verse = bookmark.verse,
                text = bookmark.text,
            ),
        )
    }

    private fun warmUpSelectedVersion(version: BibleVersion) {
        viewModelScope.launch(Dispatchers.IO) {
            repository.warmUpVersion(version)
        }
    }

    private fun saveProgress(
        version: BibleVersion?,
        comparisonVersion: BibleVersion?,
        bookIndex: Int,
        chapter: Int,
    ) {
        progressPreferences.setProgress(
            ReadingProgress(
                versionCode = version?.code,
                comparisonVersionCode = comparisonVersion?.code,
                hasSavedComparisonVersion = true,
                bookIndex = bookIndex,
                chapter = chapter,
            ),
        )
    }
}

File: ios/Sources/Presentation/Features/Reader/ReaderViewModel.swift
import Foundation

@MainActor
final class ReaderViewModel: ObservableObject {
    @Published var versions: [BibleVersion] = []
    @Published var selectedVersion: BibleVersion?
    @Published var comparisonVersion: BibleVersion?
    @Published var bookIndex = 0
    @Published var chapter = 1
    @Published var focusVerse: Int?
    @Published var verses: [BibleVerse] = []
    @Published var comparisonVerses: [BibleVerse] = []
    @Published var isLoading = true
    @Published var loadingMessage = "성경 데이터를 준비하는 중입니다."
    @Published var message: String?

    private var repository: BibleRepository?
    private var progressStore: ReadingProgressStore?
    private var dataFolderStore: DataFolderStore?
    private var dataRoot: URL?
    private var loadGeneration = 0

    func configure(environment: AppEnvironment) {
        repository = environment.repository
        progressStore = environment.progressStore
        dataFolderStore = environment.dataFolderStore
    }

    func refresh(environment: AppEnvironment) async {
        configure(environment: environment)

        guard environment.isBootstrapped else { return }

        if let bootstrapError = environment.bootstrapError {
            isLoading = false
            message = "앱 초기화에 실패했습니다. \(bootstrapError)"
            return
        }

        guard let repository, let dataFolderStore else {
            isLoading = false
            message = "성경 데이터를 불러올 준비가 되지 않았습니다."
            return
        }

        dataRoot = environment.dataFolderURL ?? dataFolderStore.defaultBibleFolder()
        guard let root = dataRoot else {
            isLoading = false
            message = "성경 데이터 폴더를 선택하거나 샘플 데이터를 다운로드해 주세요."
            versions = []
            selectedVersion = nil
            verses = []
            return
        }

        isLoading = true
        loadingMessage = "데이터 폴더에서 BDF/LFA 역본을 찾는 중입니다."
        message = nil

        let scoped = dataFolderStore.startAccessing() ?? root
        defer { dataFolderStore.stopAccessing(scoped) }

        do {
            let scanned = try await repository.scanVersions(at: scoped)
            versions = scanned
            let saved = progressStore?.load() ?? ReadingProgress()
            bookIndex = min(max(saved.bookIndex, 0), BibleCatalog.books.count - 1)
            let book = BibleCatalog.book(at: bookIndex)
            chapter = min(max(saved.chapter, 1), book.chapterCount)

            selectedVersion = saved.versionCode.flatMap { code in
                scanned.first { $0.code.caseInsensitiveCompare(code) == .orderedSame }
            } ?? scanned.first
            let defaults = UserDefaults.standard
            if defaults.bool(forKey: Self.comparisonVersionSavedKey) {
                comparisonVersion = defaults.string(forKey: Self.comparisonVersionCodeKey).flatMap { code in
                    scanned.first {
                        $0.code.caseInsensitiveCompare(code) == .orderedSame && $0.id != selectedVersion?.id
                    }
                }
            } else {
                comparisonVersion = scanned.first { $0.id != selectedVersion?.id }
            }

            loadingMessage = selectedVersion.map { "\($0.displayName) \(book.koreanName) \(chapter)장을 여는 중입니다." }
                ?? "사용 가능한 역본을 확인하는 중입니다."

            if scanned.isEmpty {
                isLoading = false
                message = "선택한 폴더에서 BDF/LFA 역본을 찾지 못했습니다."
                verses = []
                return
            }

            await reloadCurrentChapter()
            isLoading = false
        } catch {
            isLoading = false
            message = error.localizedDescription
        }
    }

    func selectVersion(_ version: BibleVersion) {
        selectedVersion = versions.first { $0.id == version.id } ?? version
        if comparisonVersion?.id == selectedVersion?.id {
            comparisonVersion = nil
        }
        message = nil
        Task { await reloadCurrentChapter() }
        saveProgress()
    }

    func selectComparison(_ version: BibleVersion?) {
        comparisonVersion = version?.id == selectedVersion?.id ? nil : version
        Task { await reloadCurrentChapter() }
        saveProgress()
    }

    func selectBook(_ index: Int, chapter newChapter: Int = 1) {
        bookIndex = index
        chapter = newChapter
        focusVerse = nil
        Task { await reloadCurrentChapter() }
        saveProgress()
    }

    func moveChapter(by delta: Int) {
        var newBookIndex = bookIndex
        var newChapter = chapter + delta
        while newChapter < 1 {
            if newBookIndex == 0 { return }
            newBookIndex -= 1
            newChapter += BibleCatalog.book(at: newBookIndex).chapterCount
        }
        while newChapter > BibleCatalog.book(at: newBookIndex).chapterCount {
            if newBookIndex >= BibleCatalog.books.count - 1 { return }
            newChapter -= BibleCatalog.book(at: newBookIndex).chapterCount
            newBookIndex += 1
        }
        bookIndex = newBookIndex
        chapter = newChapter
        focusVerse = 1
        message = nil
        Task { await reloadCurrentChapter() }
        saveProgress()
    }

    func openSearchResult(_ verse: BibleVerse) {
        if let version = versions.first(where: { $0.code.caseInsensitiveCompare(verse.versionCode) == .orderedSame }) {
            selectedVersion = version
        }
        bookIndex = verse.bookIndex
        chapter = verse.chapter
        focusVerse = verse.verse
        Task { await reloadCurrentChapter() }
        saveProgress()
    }

    func openBookmark(_ bookmark: VerseBookmark) {
        openSearchResult(BibleVerse(
            versionCode: bookmark.versionCode,
            bookIndex: bookmark.bookIndex,
            chapter: bookmark.chapter,
            verse: bookmark.verse,
            text: bookmark.text
        ))
    }

    var currentBook: BibleBook { BibleCatalog.book(at: bookIndex) }

    private func reloadCurrentChapter() async {
        loadGeneration += 1
        let generation = loadGeneration
        await loadChapter(expectedGeneration: generation)
    }

    private func resolvedDataRoot(using dataFolderStore: DataFolderStore) -> URL? {
        dataRoot ?? dataFolderStore.resolvedFolderURL() ?? dataFolderStore.defaultBibleFolder()
    }

    private func loadChapter(expectedGeneration: Int) async {
        guard expectedGeneration == loadGeneration else { return }
        guard let repository, let dataFolderStore, let selectedVersion else {
            verses = []
            comparisonVerses = []
            return
        }
        guard let root = resolvedDataRoot(using: dataFolderStore) else {
            verses = []
            comparisonVerses = []
            return
        }

        let scoped = dataFolderStore.startAccessing() ?? root
        defer { dataFolderStore.stopAccessing(scoped) }

        let book = currentBook
        let version = selectedVersion.withRoot(scoped)
        let comparison = comparisonVersion?.withRoot(scoped)
        let targetBookIndex = bookIndex
        let targetChapter = chapter

        do {
            let loadedVerses = try await repository.readChapter(version: version, book: book, chapter: targetChapter)
            guard expectedGeneration == loadGeneration else { return }
            guard bookIndex == targetBookIndex, chapter == targetChapter else { return }

            verses = loadedVerses
            if let comparison {
                comparisonVerses = try await repository.readChapter(
                    version: comparison,
                    book: book,
                    chapter: targetChapter
                )
            } else {
                comparisonVerses = []
            }
        } catch {
            guard expectedGeneration == loadGeneration else { return }
            message = error.localizedDescription
        }
    }

    private func saveProgress() {
        progressStore?.save(ReadingProgress(
            versionCode: selectedVersion?.code,
            bookIndex: bookIndex,
            chapter: chapter
        ))
        let defaults = UserDefaults.standard
        defaults.set(comparisonVersion?.code, forKey: Self.comparisonVersionCodeKey)
        defaults.set(true, forKey: Self.comparisonVersionSavedKey)
    }

    private static let comparisonVersionCodeKey = "last_comparison_version_code"
    private static let comparisonVersionSavedKey = "has_saved_comparison_version"
}
